store/postgres/src/entities.rs

```diff
@@ -310,37 +320,44 @@ pub(crate) fn make_storage_cache() -> StorageCache {
     Mutex::new(HashMap::new())
 }
 
-/// A connection into the database to handle entities which caches the
-/// layout to actual database tables. Instances of this struct must not be
-/// cached across transactions as there is no mechanism in place to notify
-/// other index nodes that a subgraph has been migrated
-pub(crate) struct Connection<'a> {
-    pub conn: PooledConnection<ConnectionManager<PgConnection>>,
-    store: &'a Store,
+/// A connection into the database to handle entities. The connection is
+/// specific to one subgraph, and can only handle entities from that subgraph
+/// or from the metadata subgraph. Attempts to access other subgraphs will
+/// generally result in a panic.
+///
+/// Instances of this struct must not be cached across transactions as there
+/// is no mechanism in place to notify other index nodes that a subgraph has
+/// been migrated
+#[derive(Constructor)]
+pub(crate) struct Connection {
+    conn: PooledConnection<ConnectionManager<PgConnection>>,
+    /// The storage of the subgraph we are dealing with; entities
+    /// go into this
+    storage: Arc<Storage>,
+    /// The layout of the subgraph of subgraphs where we keep subgraph
+    /// metadata
+    metadata: Arc<Storage>,
 }
 
-impl<'a> Connection<'a> {
-    pub(crate) fn new(
-        conn: PooledConnection<ConnectionManager<PgConnection>>,
-        store: &'a Store,
-    ) -> Connection<'a> {
-        Connection { conn, store }
-    }
-
+impl Connection {
     /// Return the storage for the subgraph. Since constructing a `Storage`
     /// object takes a bit of computation, we cache storage objects that do
     /// not have a pending migration in the Store, i.e., for the lifetime of
     /// the Store. Storage objects with a pending migration can not be
     /// cached for longer than a transaction since they might change
     /// without us knowing
-    fn storage(&self, subgraph: &SubgraphDeploymentId) -> Result<Arc<Storage>, StoreError> {
-        if let Some(storage) = self.store.storage_cache.lock().unwrap().get(subgraph) {
+    pub(crate) fn storage(
+        conn: &PgConnection,
+        store: &Store,
+        subgraph: &SubgraphDeploymentId,
+    ) -> Result<Arc<Storage>, StoreError> {
```

Comment:
I think this will now make more sense as a method on `Store`.

Reply:
Good point. Commit c28ad232